Text Resizer

jivey opened the pull request:
For: https://github.com/openstax/unified/issues/1676

jivey committed:
add text resizer

jivey committed:
update tests

jivey committed:
fix style lint

jivey committed:
use a mixin for font size calc

jivey committed:
always apply content-text-scale

jivey committed:
fix css alignment issues and use intl messages

jivey committed:
update snapshots

jivey committed:
fix indentation

jivey committed:
add specs for range input

jivey committed:
fix button alignment

jivey committed:
update snapshots

jivey committed:
Merge branch 'master' into text-resizer

jivey committed:
fix merge issues with topbar

jivey committed:
fix snapshot

jivey committed:
fix prerender test

jivey committed:
remove unused check

jivey committed:
remove branch

jivey committed:
inline window check

jivey committed:
fix lint

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
add placeholder translations

jivey commented on src/app/messages/es/messages.json:
```diff
@@ -210,5 +210,9 @@
+  "i18n:toolbar:textresizer:button:aria-label": "Cambiar el tamaño del texto",
+  "i18n:toolbar:textresizer:button:decrease:aria-label": "Disminuir el tamaño del texto",
+  "i18n:toolbar:textresizer:button:increase:aria-label": "Aumentar el tamaño del texto",
+  "i18n:toolbar:textresizer:popup:heading": "Tamaño del texto"
```
@bethshook added some placeholder translations, hoping google didn't steer me too far off

bethshook replied:
Looks good!

jivey commented on src/app/messages/pl/messages.json:
```diff
@@ -210,5 +210,9 @@
+  "i18n:toolbar:textresizer:button:aria-label": "Zmień rozmiar tekstu",
+  "i18n:toolbar:textresizer:button:decrease:aria-label": "Zmniejsz rozmiar tekstu",
+  "i18n:toolbar:textresizer:button:increase:aria-label": "Zwiększ rozmiar tekstu",
+  "i18n:toolbar:textresizer:popup:heading": "Rozmiar tekstu"
```
@firadecki 👋 Would you have some time to look at these and let me know if they work? I took them from google translate so maybe they need some changes.

firadecki replied:
@jivey Translation looks fine, grammarly correct. There is more than one translation possible for some positions but it's hard to specify without seeing text resizer. Translation is correct and good for now, I'll check if anything can be perfected during testing.

bethshook commented on generic-styles/index.less:
```diff
@@ -38,8 +68,8 @@ body {
-  font-size: 1.6rem;
-  line-height: 2.3rem;
+  .content-font-size(1.6rem);
+  line-height: calc(2.3rem * var(--content-text-scale));
```
Another place I'm noticing existing issues that will be exacerbated by large text: the [writing guide handbook](https://openstax.org/books/writing-guide/pages/18-5-writing-process-create-a-multimodal-advocacy-project). Something (tables?) is causing the content to overflow its container and it gets worse as you increase font size. But maybe both of these things should just be separate bug cards. I will approve and leave it to you.

jivey replied:
@victoria-cooke @bethshook I updated the line-heights for h2s and h3s, picked an arbitrary 125% of the font-size, could see wanting to tweak it some more, maybe slightly tighter for h2? Let me know your thoughts and I can update them:

Before | After
--- | ---
![image](https://user-images.githubusercontent.com/34174/171948408-7c8d1c69-de4a-4b68-a7c2-adc638ed067b.png) | ![image](https://user-images.githubusercontent.com/34174/171948309-0be61a61-d609-46b5-97b3-ef9043e833a8.png)

Before | After
--- | ---
![image](https://user-images.githubusercontent.com/34174/171948453-5e608cd5-5eec-4716-a265-127c01fbbb44.png) | ![image](https://user-images.githubusercontent.com/34174/171948559-912778f4-f3d7-4523-84fa-502fc46206fc.png)

jivey replied:
I forgot to label these images better: top is the default font size, bottom is the largest.

jivey replied:
And here's a single line version at default font size that helps show the slight space change, for example an h2 would get around 10px more visible space between the header and paragraph. Might be a lot easier to test on staging and adjust in a future release?

Before | After
--- | ---
![image](https://user-images.githubusercontent.com/34174/171955893-a11a84a3-8d20-434b-a44c-d73104d7c8c8.png) | ![image](https://user-images.githubusercontent.com/34174/171955778-9768d5a7-2611-42eb-8307-e5d8d58b14f1.png) |

bethshook replied:
I defer to @victoria-cooke

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
add event for text resizer dropdown

jivey committed:
fix caption size becoming too big

it uses an em, so it already is getting scaled via the container size

jivey committed:
add base line heights for h2 and h3

jivey committed:
update snapshots

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
update scroll tops with new line heights

jivey committed:
Merge branch 'text-resizer' of github.com:openstax/rex-web into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
hoist data attribue var setting to apply to both rex and cnx styles

If using dynamic styles, the .less rules for the css vars will not be
used, and we can't just set them in cnx-recipes because the cascade
for dynamic styles already starts at #main-content and we need to
check a data-attribute on that element. Moving them to the MainContent
component will ensure they are attached in both scenarios.

jivey committed:
satisfy the linter

jivey commented:
@bethshook Made one last change that might be significant, fixes a bug with using dynamic styles by ensuring the data-attribute rules always get applied: https://github.com/openstax/rex-web/pull/1598/commits/ed666bf274c6be30fdc65e9225d85c5ee551584b

jivey committed:
update snapshots

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

TomWoodward commented on src/app/content/hooks/storeTextSize.ts:
```diff
@@ -0,0 +1,14 @@
+import { setTextSize } from '../actions';
+import { textSize } from '../selectors';
+
+const hookBody: ActionHookBody<typeof setTextSize> = (services) => async() => {
```
you can add a body to the first function level here, which is run when the app is wired up and dispatch the initial `setTextSize` there, that way everything would all be in the same place

jivey replied:
I tried that but redux complains about dispatching while middleware is constructing, maybe I set it up wrong though

TomWoodward replied:
No you're probably right, you might need another hook in this file listening to locationChange for the setup

TomWoodward commented on src/app/content/reducer.ts:
```diff
@@ -18,6 +18,8 @@ import searchReducer, {initialState as initialSearchState } from './search/reduc
 import studyGuidesReducer, {initialState as initialStudyGuidesState } from './studyGuides/reducer';
 import { State } from './types';
 
+const storedTextSize = typeof window !== 'undefined' && window.localStorage.getItem('textSize');
```
you may need to have this here if my other suggestion causes a flash of the stuff at a different size, but even then the key should at least be stored in a constant

TomWoodward commented on src/app/content/components/Topbar/index.tsx:
```diff
@@ -121,18 +151,29 @@ class Topbar extends React.Component<Props, State> {
+          onDecreaseTextSize={onDecreaseTextSize}
+          onIncreaseTextSize={onIncreaseTextSize}
+          textSize={this.props.textSize}
+          onToggle={this.props.clearSearch}
```
i don't think we want to do this

jivey committed:
simplify setting css var

jivey committed:
don't close search and fix overlay issues

jivey committed:
add clarifying comment

jivey commented on src/app/content/components/popUp/Filters.tsx:
```diff
@@ -76,23 +80,31 @@ const Toggle = styled(React.forwardRef<HTMLButtonElement, ToggleProps>(
+  >
+    {children}
+  </Dropdown>
+);
```
good call

jivey committed:
use const for storage key

jivey committed:
fix lint

jivey committed:
match guard conventions

TomWoodward commented on src/app/content/components/Content.browserspec.tsx:
```diff
@@ -9,8 +9,8 @@ const TEST_CASES: { [testCase: string]: (target: Page) => Promise<void> } = {
-  Desktop: [242, 90, 122, 242, 365, 668, 761, 1263, 1607],
-  Mobile: [239, 66, 96, 239, 523, 1263, 1402, 1751, 2118],
+  Desktop: [249, 90, 122, 249, 372, 675, 768, 1270, 1614],
+  Mobile: [246, 66, 96, 246, 530, 1270, 1409, 1758, 2125],
```
Ok that makes sense, did Victoria ever get back to you? You may want to check with Larissa too.

Also did you check the math based on the number of headings in the test page to make sure you're setting the expected values?

jivey replied:
After another round of line height updates, these do not change.

jivey committed:
update snapshots

jivey committed:
fix tickmarks and button heights

jivey committed:
update line heights for h2 and h3

updating line heights after feedback from UX

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
fix search field alignment by matching content pane rules

jivey committed:
improve accessibility and button padding

jivey committed:
update snapshots

jivey committed:
revert line heights after h3 update

jivey committed:
textSize should be present via initialstate

jivey committed:
fix snapshots

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

staxly[bot] committed:
Merge branch 'master' into text-resizer

jivey committed:
move localstorage load to hook

jivey committed:
fix lint issues

jivey committed:
add another default to store for page render

jivey committed:
DRY up styles

jivey committed:
fix lint

jivey committed:
fix constant lookup

TomWoodward commented on generic-styles/index.less:
```diff
@@ -143,14 +151,16 @@ body {
   h3 {
-    font-size: 2.4rem;
+    .content-font-size(2.4rem);
+    line-height: calc(2.3rem * var(--content-text-scale));
```
you wanna make a similar helper for line height?

jivey replied:
yeah good idea, at first there were only 2 entries and didn't think about it after we updated a few other places

TomWoodward commented on src/app/content/actions.ts:
```diff
@@ -32,3 +32,5 @@ export const receivePageNotFoundId = createStandardAction('Content/receivePageNo
 export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
 export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
+
+export const setTextSize = createStandardAction('Content/setTextSize')<number>();
```
you should type this and `textResizerValueMap` with a union of const values, or replace `textResizerValueMap` with some kind of math function, otherwise you're setting yourself up to have invalid numbers go in the top and `undefined` printed in the styles

jivey replied:
Added better typing here with `as const`, does require a few `as TextResizerValue` because of the narrowing when using `.includes` but overall seems to work nicely

jivey replied:
Ended up going with some minor duplication to explicitly type the values https://github.com/openstax/rex-web/pull/1598/files#diff-51c9e59c9ff2ef74713ad6f6f68e468bc12c90fd9778beab1fa3f710c2733553R24 but maybe there is a better way, I thought a union approach wouldn't narrow it enough? Let me know if you have a suggestion!

TomWoodward commented on src/app/content/components/Topbar/index.tsx:
```diff
@@ -85,28 +91,52 @@ class Topbar extends React.Component<Props, State> {
 
     return <Styled.TopBarWrapper data-testid='topbar'>
+      {/* mobileMedium is smaller (800px) than mobile (1200px) */}
+      {/* Only hides on .mobile breakpoint, shows on desktop and .mobileMedium */}
```
what hides on mobile? does this comment conflict with the one on line 175?

jivey replied:
Styled.SearchPrintWrapper below, but these were more to keep it straight for myself while working on it, might just remove them because they feel prone to being out of date soon

TomWoodward commented on src/app/content/hooks/storeTextSize.ts:
```diff
@@ -0,0 +1,37 @@
+    value = parseInt(storedTextSize, 10);
+  }
+
+  dispatch(setTextSize(value));
```
you wanna check that value is not NaN

jivey replied:
yeah this really did need better NaN and bounds checking, added that here and also in the event callback

TomWoodward commented on src/app/content/hooks/locationChange/index.ts:
```diff
@@ -24,6 +25,7 @@ const hookBody: RouteHookBody<typeof content> = (services) => {
       loadStudyGuides(services)(),
       loadPracticeQuestions(services)(),
       initializeIntl(services)(),
+      loadStoredTextSize(services)(),
```
you only have to do this here if your hook relies on the book data being loaded, i think in your case you could compose the hook directly in `src/app/content/hooks/index.ts`

jivey committed:
add a helper for line-height as well

jivey committed:
strongly type setTextSize, add parsed value boundary checking

jivey committed:
fix indentation, add fallback margin in case text resizer can't show

jivey committed:
move hook to live with other content hooks

jivey committed:
remove comments

jivey committed:
🤔

jivey committed:
move TextResizer constants, substitute 'as const'

jivey committed:
fix import order

jivey commented on src/app/content/constants.ts:
```diff
@@ -20,3 +20,12 @@ export const highlightStyles: Array<{label: HighlightColorEnum, passive: string,
 ];
+
+export type TextResizerValue = -2 | -1 | 0 | 1 | 2 | 3;
+export const textResizerValues: TextResizerValue[] = [-2, -1, 0, 1, 2, 3];
```
No cause `as const` doesn't get transformed, babel thinks it's a typo.

jivey commented on src/app/content/components/Page/connector.ts:
```diff
@@ -30,6 +30,7 @@ export interface PagePropTypes {
   services: AppServices & MiddlewareAPI;
   addToast: typeof addToast;
   systemQueryParams: SystemQueryParams;
+  textSize: number;
```
Oops I shouldn't not have relied on tslint complaining to me about these, good catch...

jivey committed:
sidebar is a number const

jivey committed:
fix content/types not getting updated to use TextResizerValue

jivey committed:
snapshots

jivey committed:
check for value presence with an explicit null check
